src/ucp/rndv/proto_rndv.c

```diff
@@ -363,28 +346,143 @@ ucp_proto_rndv_ctrl_init(const ucp_proto_rndv_ctrl_init_params_t *params,
                                                 min_length,
                                                 range_max_length, SIZE_MAX,
                                                 params->perf_bias,
                                                 parallel_stages, 2);
         if (status != UCS_OK) {
-            goto out_deref_perf_node;
+            return status;
         }
 
         ucp_proto_perf_node_deref(&remote_perf.node);
 
-        min_length = range_max_length - 1;
+        min_length = range_max_length + 1;
     } while ((remote_range++)->max_length < max_length);
 
-    status = UCS_OK;
+    return UCS_OK;
+}
+
+void
+ucp_proto_rndv_set_variant_config(const ucp_proto_init_params_t *init_params,
+                                  const ucp_proto_init_elem_t *proto,
+                                  const ucp_proto_select_param_t *select_param,
+                                  void *priv, ucp_proto_config_t *cfg)
+{
+    cfg->priv           = priv;
+    cfg->cfg_thresh     = proto->cfg_thresh;
+    cfg->ep_cfg_index   = init_params->ep_cfg_index;
+    cfg->proto          = ucp_protocols[proto->proto_id];
+    cfg->rkey_cfg_index = init_params->rkey_cfg_index;
+    cfg->select_param   = *select_param;
+}
+
+void ucp_proto_rndv_ctrl_probe(const ucp_proto_rndv_ctrl_init_params_t *params,
+                               ucp_lane_index_t lane)
+{
+    const ucp_proto_init_params_t *init_params = &params->super.super;
+    ucp_proto_rndv_ctrl_priv_t *rpriv          = init_params->priv;
+    const ucp_proto_select_init_protocols_t *remote_proto_init;
+    ucp_proto_perf_range_t ctrl_perf, *last_range;
+    ucp_proto_select_param_t remote_select_param;
+    ucp_proto_select_elem_t *remote_proto_select;
+    ucp_proto_init_elem_t *remote_proto;
+    ucp_proto_caps_t *remote_caps;
+    ucs_status_t status;
+    size_t min_length, max_length, cfg_thresh, cfg_priority;
+    void *remote_priv;
+
+    ucs_assert(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE);
+    ucs_assert(!(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG));
+
+    if (!ucp_proto_common_init_check_err_handling(&params->super)) {
+        status = UCS_ERR_UNSUPPORTED;
+        goto out;
+    }
+
+    /* Initialize 'rpriv' structure */
+    ucp_proto_rndv_ctrl_init_priv(params, lane);
+
+    /* Init remote proto */
+    remote_select_param = ucp_proto_rndv_remote_select_param_init(params);
+    status              = ucp_proto_rndv_ctrl_select_remote_proto(
+            params, &remote_select_param, rpriv->md_map, &remote_proto_select);
+    if (status != UCS_OK) {
+        goto out;
+    }
+    remote_proto_init = &remote_proto_select->proto_init;
+
+    /* Add variants for each remote proto */
+    ucs_array_for_each(remote_proto, &remote_proto_init->protocols) {
+        remote_caps = &remote_proto->caps;
+        /* Skip empty variants and reconfig proto */
+        if ((remote_caps->num_ranges == 0) || (remote_proto->proto_id == 0)) {
```

Comment:
can reconfig be the only option?

Reply:
In that case there are no variant added so the protocol would be unavailable.